File: cakk-api/src/test/java/com/cakk/api/provider/JwtProviderImplTest.java
package com.cakk.api.provider;

import static com.cakk.api.common.utils.TestUtils.*;
import static com.cakk.common.enums.Role.*;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Import;
import org.springframework.security.core.Authentication;
import org.springframework.test.util.ReflectionTestUtils;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;

import com.cakk.api.common.annotation.TestWithDisplayName;
import com.cakk.api.common.base.MockitoTest;
import com.cakk.api.config.JwtConfig;
import com.cakk.api.provider.jwt.JwtProviderImpl;
import com.cakk.core.vo.JsonWebToken;
import com.cakk.domain.mysql.entity.user.User;
import com.cakk.domain.redis.repository.TokenRedisRepository;

@Import(JwtConfig.class)
@DisplayName("Jwt Provider 테스트")
class JwtProviderImplTest extends MockitoTest {

	private JwtProviderImpl jwtProviderImpl;

	@Mock
	private TokenRedisRepository tokenRedisRepository;

	@Value("${jwt.secret}")
	private String secretKey;

	@BeforeEach
	void setup() {
		jwtProviderImpl = new JwtProviderImpl(
			Keys.hmacShaKeyFor(Decoders.BASE64.decode(SECRET_KEY)),
			tokenRedisRepository,
			ACCESS_TOKEN_EXPIRED_SECOND,
			REFRESH_TOKEN_EXPIRED_SECOND,
			GRANT_TYPE,
			USER_KEY
		);
	}

	@TestWithDisplayName("토큰 생성에 성공한다.")
	void generateToken1() {
		// given
		User user = getConstructorMonkey().giveMeOne(User.class);

		// when
		JsonWebToken jwt = jwtProviderImpl.generateToken(user);

		// then
		Assertions.assertNotNull(jwt.getAccessToken());
		Assertions.assertNotNull(jwt.getRefreshToken());
		Assertions.assertNotNull(jwt.getGrantType());
	}

	@TestWithDisplayName("액세스 토큰으로부터 인증 정보를 가져온다.")
	void getAuthentication() {
		// given
		User user = getConstructorMonkey().giveMeOne(User.class);
		ReflectionTestUtils.setField(user, "id", 1L);
		ReflectionTestUtils.setField(user, "role", USER);

		String accessToken = jwtProviderImpl.generateToken(user).getAccessToken();

		// when
		Authentication authentication = jwtProviderImpl.getAuthentication(accessToken);

		// then
		Assertions.assertEquals(authentication.getAuthorities().toString(), "[%s]".formatted(USER.getSecurityRole()));
	}

	@TestWithDisplayName("토큰으로부터 Claim 정보를 가져온다.")
	void parseClaims() {
		// given
		User user = getConstructorMonkey().giveMeOne(User.class);
		ReflectionTestUtils.setField(user, "id", 1L);
		ReflectionTestUtils.setField(user, "role", USER);

		String accessToken = jwtProviderImpl.generateToken(user).getAccessToken();

		// when
		Claims claims = jwtProviderImpl.parseClaims(accessToken);

		// then
		Assertions.assertNotNull(claims);
	}
}

File: cakk-api/src/test/java/com/cakk/api/provider/JwtProviderImplTest.kt
package com.cakk.api.provider

import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.mockito.Mock
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.Authentication
import org.springframework.test.util.ReflectionTestUtils

import com.cakk.api.common.annotation.TestWithDisplayName
import com.cakk.api.common.base.MockitoTest
import com.cakk.api.common.fixture.FixtureCommon.fixtureMonkey
import com.cakk.api.common.utils.*
import com.cakk.api.provider.jwt.JwtProviderImpl
import com.cakk.common.enums.Role
import com.cakk.core.vo.JsonWebToken
import com.cakk.infrastructure.cache.repository.TokenRedisRepository
import com.cakk.infrastructure.persistence.entity.user.UserEntity

@DisplayName("Jwt Provider 테스트")
internal class JwtProviderImplTest : MockitoTest() {

	private lateinit var jwtProviderImpl: JwtProviderImpl

	@Mock
	private lateinit var tokenRedisRepository: TokenRedisRepository

	@Value("\${jwt.secret}")
	private val secretKey: String = SECRET_KEY

	@BeforeEach
	fun setup() {
		jwtProviderImpl = JwtProviderImpl(
			Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKey)),
			tokenRedisRepository,
			ACCESS_TOKEN_EXPIRED_SECOND,
			REFRESH_TOKEN_EXPIRED_SECOND,
			GRANT_TYPE,
			USER_KEY
		)
	}

	@TestWithDisplayName("토큰 생성에 성공한다.")
	fun generateToken1() {
		// given
		val userEntity = fixtureMonkey.giveMeOne(UserEntity::class.java)

		// when
		val jwt: JsonWebToken = jwtProviderImpl.generateToken(userEntity)

		// then
		jwt.accessToken shouldNotBe null
		jwt.refreshToken shouldNotBe null
		jwt.grantType shouldNotBe null
	}

	@TestWithDisplayName("액세스 토큰으로부터 인증 정보를 가져온다.")
	fun getAuthentication() {
		// given
		val userEntity = fixtureMonkey.giveMeOne(UserEntity::class.java)
		ReflectionTestUtils.setField(userEntity, "id", 1L)
		ReflectionTestUtils.setField(userEntity, "role", Role.USER)

		val accessToken: String = jwtProviderImpl.generateToken(userEntity).accessToken

		// when
		val authentication: Authentication = jwtProviderImpl.getAuthentication(accessToken)

		// then
		authentication.authorities.toString() shouldBe "[${Role.USER.securityRole}]"
	}

	@TestWithDisplayName("토큰으로부터 Claim 정보를 가져온다.")
	fun parseClaims() {
		// given
		val userEntity = fixtureMonkey.giveMeOne(UserEntity::class.java)
		ReflectionTestUtils.setField(userEntity, "id", 1L)
		ReflectionTestUtils.setField(userEntity, "role", Role.USER)

		val accessToken: String = jwtProviderImpl.generateToken(userEntity).accessToken

		// when
		val claims = jwtProviderImpl.parseClaims(accessToken)

		// then
		claims shouldNotBe null
	}
}
